examples/wave.cxx

```diff
@@ -0,0 +1,372 @@
+    
+    // loop over points
+    for(int ty=0; ty<nx; ++ty) {
+      for(int tx=0; tx<nx; ++tx) {
+        
+        const int id = tx + ty*nx;
+        T P_old  = P1[id]; 
+        T P_curr = P2[id]; 
+        
+        //Compute laplacian
+        T lap = 0.0;
+        
+        for(int r=-sr; r<=sr; ++r) {
+          const int xi  = (tx+r+nx)%nx;
+          const int idx = xi + nx*ty;
+          lap += coeff[r+sr]*P2[idx]; 
+          
+          const int yi  = (ty+r+nx)%nx;
+          const int idy = tx + nx*yi;
+          lap += coeff[r+sr]*P2[idy];
+        }
+        
+        //finite difference stencil
+        P1[id] = 2*P_curr - P_old + ct*lap;
+      }
+    }
+    
+  }
+
+};
+
+// RAJA implementation for CUDA
+template <size_t BLOCK_SIZE>
+struct Wave<RAJA::cuda_exec<BLOCK_SIZE>, RAJA::cuda_exec<BLOCK_SIZE> > {
+  template <typename T>
+    void operator()(T* P1, T* P2, int nx, double ct) {
+
+    RAJA::forallN<RAJA::NestedPolicy<
+    RAJA::ExecList<
+    RAJA::cuda_threadblock_y_exec<BLOCK_SIZE>,
+    RAJA::cuda_threadblock_x_exec<BLOCK_SIZE> >>>(
+    RAJA::RangeSegment(0, nx), RAJA::RangeSegment(0, nx), [=] __device__ (int tidy, int tidx) {
+                                             
+    double coeff[5] = {-1.0/12.0,4.0/3.0,-5.0/2.0,4.0/3.0,-1.0/12.0}; //fourth order scheme
+    
+    // loop over points        
+    const int id = tidx + tidy*nx;
+    T P_old  = P1[id]; 
+    T P_curr = P2[id]; 
+    
+    //Compute laplacian
+    T lap = 0.0;
+    
+    for(int r=-sr; r<=sr; ++r) {
+      const int xi  = (tidx+r+nx)%nx;
+      const int idx = xi + nx*tidy;
+      lap += coeff[r+sr]*P2[idx]; 
+      
+      const int yi  = (tidy+r+nx)%nx;
+      const int idy = tidx + nx*yi;
+      lap += coeff[r+sr]*P2[idy];
+    }
+    
+    //finite difference stencil
+    P1[id] = 2*P_curr - P_old + ct*lap;
+          
+  }); 
+    
+
+  }
+};
+
+
+double wave_sol(double t, double x, double y);
+void set_ic(double *P1, double *P2, double t0, double t1, grid_s grid);
+void compute_err(double *P, double tf, grid_s grid);
+
+/*
+  Example 4: Two dimensional solver for the acoustic wave eqation P_tt = cc(P_xx + P_yy)  
+
+  ------[Details]----------------------
+  Scheme uses a second order temporal finite diffrence discretization and a
+  fourth order spatial discretization in space. Periodic boundary conditions are assumed.
+  NOTE: The x and y dimensions are discretized identically. 
+
+*/
+int main(int RAJA_UNUSED_ARG(argc), char** RAJA_UNUSED_ARG(argv[]))
+{
+
+  std::cout<<"Example 4. Time-Domain Finite Difference Solver For The Acoustic Wave Equation"<<std::endl;
+  
+  //Wave speed squared
+  double cc = 1./2.0;
+
+  //Multiplier for spatial discretization
+  double factor = 8.0;
+
+  //Discretization of the domain. 
+  //The same discretization of the x-dimension wil be used for the y-dimension
+  grid_s grid; 
+  grid.ox   = -1; 
+  grid.dx   = 0.1250/factor;
```

Comment:
They could be anything. ox corresponds to the origin of the x-component, dx is the grid spacing (I assume equidistant points), and nx is the total number of grid points. My solver assumes a periodic solution though.... so we can somewhat vary the details.